clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,689 @@
+{-|
+  Copyright   :  (C) 2024, Rowan Goemans <goemansrowan@gmail.com>
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE DeriveLift #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+{-# OPTIONS_HADDOCK hide #-}
+
+module Clash.Cores.Crc.Internal where
+
+import           Clash.Prelude
+import           Clash.Class.HasDomain       (TryDomain, TryDomainResult(NotFound))
+import qualified Clash.Sized.Vector as V     (toList)
+
+import qualified Data.List as L
+import           Data.Maybe
+import           Data.Proxy                  (Proxy(..))
+import           Unsafe.Coerce               (unsafeCoerce)
+
+import qualified Language.Haskell.TH as TH
+import qualified Language.Haskell.TH.Syntax as TH
+import           Type.Reflection
+
+-- | Contains all the parameters to generate a CRC implementation
+data CrcParams (crcWidth :: Nat) (dataWidth :: Nat)
+  = CrcParams
+      { _crcWidth         :: SNat crcWidth
+      -- ^ Bit width of CRC word. Also known as "width" in the Williams model.
+      --   See http://techref.massmind.org/techref/method/math/crcguide.html
+      , _crcDataWidth     :: SNat dataWidth
```

Comment:
I think this is actually an implementation detail, and should thus go into `SoftwareCRC` and `HardwareCRC` types.

I was also not a particular fan of using:
* `EmptyDataType`
* Proxy arguments
* Type Applications

To pass in a particular crc configuration. I've thus created an slightly altered version that uses `newtype` instead over at: https://github.com/clash-lang/clash-compiler/commit/03c3e3acf315b76288d1f157c88357dba60a787a

Let me know your thoughts. Also @DigitalBrains1 I'd like to hear your thoughts on this.

Reply:
I do realize you lose canonicity as in principle you could make multiple configurations that would e.g. all be of type `Crc3_gsm`. Meaning that in principle you could end up with one part _encoding_ with a certain `Crc3_gsm` configuration, and another part _decoding_ with a __different__ `Crc3_gsm` configuration. But I don't know how important we think it is to protect ourselves against that vs the aesthetics/usability of the library.

Reply:
Perhaps we could simplify it even further compared to https://github.com/clash-lang/clash-compiler/commit/03c3e3acf315b76288d1f157c88357dba60a787a and even drop the newtype wrapper and just have:

```
crc3_gsm :: CrcParams 3
```

Reply:
I think that is too far. I expect users would like to write circuits polymorph in datawidth/lanes but with a fixed CRC.

Something like 

```
foo :: HardwareCrc Crc3_gsm dataWidth nLanes => Crc3_gsm -> ...
```

Then the user of `foo` is responsible for deriving the appropriate `HardwareCrc`. If you don't have this newtype there is no way to communicate and enforce this.

Reply:
We're defining unique CRC's at the type level, with `deriveHardwareCrc`. This feels like that if there is a term-level representation of such a type, it should be a singleton? What's the purpose of the `crc3_gsm` term, what does it communicate?

If  instead you go the term-route, you could also construct these terms with Template Haskell, and make the term completely contain the `CrcHardwareParams`. No further data at the type level, just make the term contain everything needed.

But I'll just go with what you two can agree on.

Reply:
> I'm also not super sure the Coercible business doesn't makes things harder to understand rather than easier.

I'd be happier if we hid all that from the user, it definitely makes understanding the stuff harder.

Reply:
Ah, also, if you base stuff on Christiaans branch, I'd like to have another go at a review. In other words, **changes requested**. :-)

Reply:
`crc3_gsm` in this case would just be a "type hint" to know for which `HardwareCrc` you want to do something. Exactly the same as `Proxy` we had before but without the `Proxy`.

Reply:
Actually my very first commit reflects this: https://github.com/clash-lang/clash-compiler/pull/2694/commits/e4cfacceda4da54f7cb0556e98a9983ed10d40a2#diff-80045d2b23ddac9d150e56d2c8b51fb3ed39401a4da0da69d6908927acad6761

Reply:
Alright, I understand now. In that case, could we move from "an empty data type + proxy" to a "unit data type"? e.g.
```haskell
data Crc3_gsm = Crc3_gsm
```